src/util/address.rs

```diff
@@ -364,6 +364,45 @@ impl Address {
         };
         format!("{}:{:#}", schema, self)
     }
+
+    /// Parsed addresses do not always have *one* network. The problem is that legacy testnet,
+    /// regtest and signet addresse use the same prefix instead of multiple different ones. When
+    /// parsing, such addresses are always assumed to be testnet addresses (the same is true for
+    /// bech32 signet addresses). So if one wants to check if an address belongs to a certain
+    /// network a simple comparison is not enough anymore. Instead this function can be used.
+    ///
+    /// ```rust
+    /// use bitcoin::{Address, Network};
+    ///
+    /// let address: Address = "2N83imGV3gPwBzKJQvWJ7cRUY2SpUyU6A5e".parse().unwrap();
+    /// assert!(address.is_valid_for_network(Network::Testnet));
+    /// assert!(address.is_valid_for_network(Network::Regtest));
+    /// assert!(address.is_valid_for_network(Network::Signet));
+    ///
+    /// assert_eq!(address.is_valid_for_network(Network::Bitcoin), false);
+    ///
+    /// let address: Address = "32iVBEu4dxkUQk9dJbZUiBiQdmypcEyJRf".parse().unwrap();
+    /// assert!(address.is_valid_for_network(Network::Bitcoin));
+    /// assert_eq!(address.is_valid_for_network(Network::Testnet), false);
+    /// ```
+    pub fn is_valid_for_network(&self, network: Network) -> bool {
+        const LEGACY_EQUIVALENCE_CLASSES: &[&[Network]] = &[
+            &[Network::Bitcoin],
+            &[Network::Testnet, Network::Regtest, Network::Signet],
+        ];
+        const SEGWIT_EQUIVALENCE_CLASSES: &[&[Network]] = &[
+            &[Network::Bitcoin],
+            &[Network::Regtest],
+            &[Network::Testnet, Network::Signet],
+        ];
+
+        let classes = match self.address_type() {
+            Some(AddressType::P2pkh) | Some(AddressType::P2sh) => LEGACY_EQUIVALENCE_CLASSES,
+            _ => SEGWIT_EQUIVALENCE_CLASSES,
```

Comment:
Is this expected behavior for undefined addresses for future segwit versions, or non 20/32 byte v0 programs?

Reply:
For me yes, idk if it _should_ be that way. My interpretation was that the prefixes are independent of the internal versioning and thus this is the correct approach for now. Of course there is still the problem that if a new address format got introduced the catch-all nature of this match arm will make it non-obvious that this piece of code needs changing. I'm not sure what to do about that. Maybe I should match over the payload instead? That could be exhaustive and thus would become a compile error if the address payload enum got extended (but wouldn't detect a new encoding for some witness version).